saleor/graphql/product/bulk_mutations/product_class_recommendation.py

```diff
@@ -0,0 +1,246 @@
+        model = models.ProductClassRecommendation
+        description = "Update product class."
+        permissions = (ProductClassPermissions.MANAGE_PRODUCT_CLASS,)
+        error_type_class = ProductClassRecommendationError
+        error_type_field = "product_class_recommendation_errors"
+
+    @classmethod
+    def config_input_cls(cls):
+        return ProductClassRecommendationBulkUpdateInput
+
+    @classmethod
+    def add_field(cls, cleaned_input, data):
+        cleaned_input["updated_by"] = data
+        return cleaned_input
+
+    @classmethod
+    @traced_atomic_transaction()
+    def perform_mutation(cls, _root, info, **data):
+        instances = cls.validate(_root, info, **data)
+        for instance in instances:
+            instance.updated_at = timezone.datetime.now()
+        models.ProductClassRecommendation.objects.bulk_update(
+            instances,
+            [
+                "listing_id",
+                "product_class_qty",
+                "product_class_value",
+                "product_class_recommendation",
+            ],
+        )
+        return ProductClassRecommendationBulkUpdate(
+            count=len(instances), product_class_recommendations=instances
+        )
+
+
+class ProductClassRecommendationBulkDelete(ModelBulkDeleteMutation):
+    class Arguments:
+        ids = graphene.List(
+            graphene.ID,
+            required=True,
+            description="List of product class IDs to delete.",
+        )
+
+    class Meta:
+        model = models.ProductClassRecommendation
+        description = "Delete product class."
+        permissions = (ProductClassPermissions.MANAGE_PRODUCT_CLASS,)
+        error_type_class = ProductClassRecommendationError
+        error_type_field = "product_class_recommendation_errors"
+
+
+class ProductClassRecommendationBulkChangeStatus(
+    BaseBulkMutation, ProductClassRecommendationChangeStatusMixin
+):
+    class Arguments:
+        ids = graphene.List(
+            graphene.NonNull(graphene.ID),
+            required=True,
+            description="Ids of product class to update status.",
+        )
+        status = ProductClassRecommendationEnum(
+            required=True,
+            description="Status of product class to update status.",
+        )
+
+    class Meta:
+        model = models.ProductClassRecommendation
+        description = "Update status product class."
+        permissions = (ProductClassPermissions.MANAGE_PRODUCT_CLASS,)
+        error_type_class = ProductClassRecommendationError
+        error_type_field = "product_class_recommendation_errors"
+
+    @classmethod
+    def clean_instance(cls, info, instance):
+        status = info.variable_values["status"]
+        cls.check_permission_change_status(info, status)
+        # cls.check_instance(instance)
+
+    @classmethod
+    def save_bulk_metadata(cls, listing_ids):
+        group_data_by_listing = {}
+        # each listing have 1 or 2 row is True
+        product_classes = (
```

Comment:
chỗ này vẫn query ra tất cả product_class cho từng listing

anh thấy em thêm vào row_number, sao em không dùng nó để filter nhỉ?

Reply:
Em cũng thử rồi ạ. Lý do là ORM django không cho phép dùng fillter với window fuc ạ. Nếu dùng thì hiển thị lỗi như này `django.db.utils.NotSupportedError: Window is disallowed in the filter clause.`  
=> Nên phải viết sql_raw kiểu subquery ạ. :)))

Reply:
sao em không dùng subquery nhỉ?

như anh làm thì chỉ cần 2 cái subquery: 1 cho current, 1 cho previous là xong
filter được mà không phải dùng window là gì cho rắc rôi

Reply:
Vâng anh. Tạm thời hôm nay em vẫn code theo cách của em ạ. Để mai em thử theo hướng của anh ạ :D